### PR TITLE
DEP: Deprecate private namespaces in ``scipy.optimize``

### PR DESCRIPTION
**Pull request by czgdp1807**
#### Reference issue

gh-14360

#### Additional information

cc: @rgommers

### CODE REVIEW

File: scipy/optimize/lbfgsb.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
+# This file is not meant for public use and will be removed in SciPy v2.0.0.
+# Use the `scipy.optimize` namespace for importing the functions
+# included below.
+
+import warnings
+from . import _lbfgsb_py
+
+
+__all__ = [  # noqa: F822    'LbfgsInvHessProduct',
```
Comment (rgommers) [inline scipy/optimize/lbfgsb.py:9]: Something is wrong here, `LbfgsInvHessProduct` should be on the next line.

**Review by rgommers — CHANGES_REQUESTED** [review]
The second commit needs a couple of changes to `setup.py` - there are still `moduleTNC.so` and `minpack2.so` left. Could you give those an underscore as well?

One other minor issue I spotted.

**Comment by rgommers** [comment]
Also, CI is still broken.

File: scipy/optimize/lbfgsb.py (hunk 1 above)
Comment (czgdp1807) [reply scipy/optimize/lbfgsb.py:9]: Ah! Yes. I think I missed out on pressing an enter here 😅.

**Comment by czgdp1807** [comment]
> there are still `moduleTNC.so` and `minpack2.so` left. Could you give those an underscore as well?

Ah! okay. No worries. I will do that.

**Comment by czgdp1807** [comment]
I am fixing mypy errors. Will push after verifying locally.

**Comment by rgommers** [comment]
The doc build is still broken, trying to use `optimize.optimize`

**Comment by czgdp1807** [comment]
I get the following warning in docs,
```python
docstring of _odrpack.odr:16: WARNING: py:obj reference target not found: ODR
docstring of _odrpack.odr:28: WARNING: py:obj reference target not found: ODR
/Users/czgdp1807/scipy_project/scipy/build/testenv/lib/python3.9/site-packages/scipy/optimize/_root.py:docstring of scipy.optimize._root.root:86: WARNING: py:obj reference target not found: nonlin
/Users/czgdp1807/scipy_project/scipy/build/testenv/lib/python3.9/site-packages/scipy/odr/__init__.py:docstring of scipy.odr:29:<autosummary>:1: WARNING: py:obj reference target not found: scipy.odr.ODR
/Users/czgdp1807/scipy_project/scipy/build/testenv/lib/python3.9/site-packages/scipy/optimize/__init__.py:docstring of scipy.optimize:396: WARNING: py:mod reference target not found: scipy.optimize.nonlin
release/0.11.0-notes.rst:85: WARNING: py:obj reference target not found: nonlin
release/1.6.0-notes.rst:47: WARNING: py:obj reference target not found: scipy.odr.ODR
```
And the following message,
```
build finished with problems, 14 warnings.
make: *** [html-scipyorg] Error 1
Traceback (most recent call last):
  File "/Users/czgdp1807/scipy_project/scipy/runtests.py", line 566, in <module>
    main(argv=sys.argv[1:])
  File "/Users/czgdp1807/scipy_project/scipy/runtests.py", line 215, in main
    subprocess.run(cmd, check=True)
  File "/Users/czgdp1807/mambaforge/envs/scipy-dev/lib/python3.9/subprocess.py", line 528, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['make', '-Cdoc', 'PYTHON="/Users/czgdp1807/mambaforge/envs/scipy-dev/bin/python"', 'html-scipyorg', 'latex', 'SPHINXOPTS="-j2"']' returned non-zero exit status 2.
```
Are they okay?

**Comment by czgdp1807** [comment]
I think other test failures are unrelated?

**Comment by rgommers** [comment]
I'll have a look, the docs warnings are slightly odd - unclear to me why the ODR ones are showing up now for example.

**Comment by czgdp1807** [comment]
I think in release notes we use module names before this deprecation. Hence, now `nonlin` doesn't exist because it's converted to `_nonlin`. Should we update release notes as well?

**Comment by rgommers** [comment]
> I think in release notes we use module names before this deprecation. Hence, now `nonlin` doesn't exist because it's converted to `_nonlin`. Should we update release notes as well?

yes we should, just to avoid doc build warnings

**Comment by rgommers** [comment]
Hmm, I fixed most things, but `nonlin` is a little annoying. While there is nothing in `optimize.nonlin` that is not also present in `optimize`, it had its own submodule docs for some reason: http://scipy.github.io/devdocs/reference/optimize.nonlin.html#module-scipy.optimize.nonlin.

I'd prefer to get rid of that, moving to `_nonlin` and merging the docs content with the main `optimize` docs. But I want to flag that and make sure we're okay with that. The alternative is to add `optimize.nonlin` to the list of public API modules in http://scipy.github.io/devdocs/reference/index.html#api-definition and revert some of the changes in this PR (and keep the duplicate exposing of functions in the two namespaces). Anyone have any objections to getting rid of the `nonlin` namespace @scipy/scipy-core-team?

**Comment by larsoner** [comment]
> Anyone have any objections to getting rid of the nonlin namespace @scipy/scipy-core-team?

None here, sounds like a good plan to me

**Review by rgommers — APPROVED** [review]

**Comment by rgommers** [comment]
Only thumbs up for not making `nonlin` public. Also, I realized after writing this that all functions in it are legacy, `root(..., method='xxx')` is recommended instead.

CI is green except for the expected linting complaints, so in it goes. Thanks @czgdp1807, all.

**Pull request merged**